### ISSUE TITLE
HTTPS ServiceEntry TLS random failures

### ISSUE THREAD

**Issue by eperdeme** [istio/istio#20477]
**Bug description**
In our Isito cluster we enable, REGISTRY_ONLY for outbound traffic, as such any outbound traffic requires a ServiceEntry.

We find connections to HTTPS endpoints randomly fail to initiate and get blackholed from our pods when specifying the protocol as HTTPS in ServiceEntry, if we specify "protocol: TCP" we do not experience this issue.

Below is the ServiceEntry used
```
kind: ServiceEntry
apiVersion: networking.istio.io/v1alpha3
metadata:
  name: git-nonprod
  namespace: nflynn
  selfLink: >-
    /apis/networking.istio.io/v1alpha3/namespaces/nflynn/serviceentries/git-nonprod
  uid: 53ea85d5-3cfb-11ea-8723-0afaefed8f04
  resourceVersion: '23230583'
  generation: 2
  creationTimestamp: '2020-01-22T09:41:13Z'
  labels:
    app.kubernetes.io/managed-by: Helm

spec:
  hosts:
    - git.nonprod.company.plc
  addresses: ~
  ports:
    - name: https
      number: 443
      protocol: HTTPS
  location: MESH_EXTERNAL
  resolution: DNS
  endpoints: ~
  exportTo: ~
  subjectAltNames: ~
```

Ran inside one of the pods
```
bash-4.2# for i in {0..10}; do curl -vs https://git.nonprod.company.plc/  2>&1 | grep ERROR; echo $i; done
0
1
2
3
* NSS error -5938 (PR_END_OF_FILE_ERROR)
4
5
6
7
8
9
10
```

Trace logs on the envoy-proxy shows the follow when it fails;
```
2020-01-24T09:26:57.569538058Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.569][32][debug][filter] [external/envoy/source/extensions/filters/listener/original_dst/original_dst.cc:18] original_dst: New connection accepted
2020-01-24T09:26:57.569569253Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.569][32][debug][filter] [external/envoy/source/extensions/filters/listener/tls_inspector/tls_inspector.cc:72] tls inspector: new connection accepted
2020-01-24T09:26:57.569587368Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.569][32][trace][filter] [external/envoy/source/extensions/filters/listener/tls_inspector/tls_inspector.cc:165] tls inspector: recv: -1
2020-01-24T09:26:57.669721879Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.669][32][debug][conn_handler] [external/envoy/source/server/connection_handler_impl.cc:179] listener filter times out after 100 ms
2020-01-24T09:26:57.669870353Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.669][32][debug][conn_handler] [external/envoy/source/server/connection_handler_impl.cc:182] fallback to default listener filter
2020-01-24T09:26:57.670011677Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.669][32][debug][filter] [src/envoy/tcp/mixer/filter.cc:30] Called tcp filter: Filter
2020-01-24T09:26:57.670071075Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.670][32][debug][filter] [src/envoy/tcp/mixer/filter.cc:40] Called tcp filter: initializeReadFilterCallbacks
2020-01-24T09:26:57.670134343Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.670][32][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:204] [C23] new tcp proxy session
2020-01-24T09:26:57.670207814Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.670][32][trace][connection] [external/envoy/source/common/network/connection_impl.cc:293] [C23] readDisable: enabled=true disable=true
2020-01-24T09:26:57.670269893Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.670][32][debug][filter] [src/envoy/tcp/mixer/filter.cc:133] [C23] Called tcp filter onNewConnection: remote 100.72.212.179:48352, local 10.210.163.136:3129
2020-01-24T09:26:57.67036408Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.670][32][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:347] [C23] Creating connection to cluster BlackHoleCluster
2020-01-24T09:26:57.670441004Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.670][32][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:1268] no healthy host for TCP connection pool
2020-01-24T09:26:57.670498294Z [Envoy (Epoch 0)] [2020-01-24 09:26:57.670][32][debug][connection] [external/envoy/source/common/network/connection_impl.cc:104] [C23] closing data_to_write=0 type
```

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
The connection works every time.

**Steps to reproduce the bug**
Deploy the ServiceEntry above with REGISTRY_ONLY enabled. Loop through a HTTPS endpoint for connection failures.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

client version: 1.4.3
citadel version: 1.4.3
galley version: 1.4.3
ingressgateway version: 1.4.3
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
nodeagent version: 
pilot version: 1.4.3
policy version: 1.4.3
sidecar-injector version: 1.4.3
telemetry version: 1.4.3
data plane version: 1.4.3 (5 proxies)

Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:16:51Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.9-eks-c0eccc", GitCommit:"c0eccca51d7500bb03b2f163dd8d534ffeb2f7a2", GitTreeState:"clean", BuildDate:"2019-12-22T23:14:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}

**How was Istio installed?**
Istioctl

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS - EKS.

**Comment by howardjohn** [comment]
Seems likely the same as #19660, can you go through the mitigation there to see if it fixes this?

**Comment by GregHanson** [comment]
@eperdeme, if you are not using protocol sniffing, you could also try `global.proxy.protocolDetectionTimeout=0s` as a deployment option.  related to this PR https://github.com/istio/istio/pull/20248

**Comment by eperdeme** [comment]
I've tried the below but it does not seem to disable sniffing;

```
--- 
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec: 
  profile: sds
  telemetry:
    enabled: true
  values:
    global:
      proxy:
        protocolDetectionTimeout: 0s
      outboundTrafficPolicy:
        mode: REGISTRY_ONLY
    pilot:
      enableProtocolSniffingForOutbound: false
      enableProtocolSniffingForInbound: false
```

The settings don't seem to take effect, I have deleted the pod still seems to be sniffing.

```2020-01-27T11:10:33.18320854Z [Envoy (Epoch 0)] [2020-01-27 11:10:33.183][30][debug][conn_handler] [external/envoy/source/server/connection_handler_impl.cc:179] listener filter times out after 100 ms
```

I'll keep poking around, I must be doing something wrong.

**Comment by eperdeme** [comment]
Sometimes it's 90% failure, sometimes 0%, there seems no logic to see.

```
bash-4.2# curl  https://dynamodb.eu-west-1.amazonaws.com -vs
* About to connect() to dynamodb.eu-west-1.amazonaws.com port 443 (#0)
*   Trying 52.94.24.50...
* Connected to dynamodb.eu-west-1.amazonaws.com (52.94.24.50) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* NSS error -5938 (PR_END_OF_FILE_ERROR)
* Encountered end of file
* Closing connection 0
bash-4.2# curl  https://dynamodb.eu-west-1.amazonaws.com -vs
* About to connect() to dynamodb.eu-west-1.amazonaws.com port 443 (#0)
*   Trying 52.94.25.174...
* Connected to dynamodb.eu-west-1.amazonaws.com (52.94.25.174) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* NSS error -5938 (PR_END_OF_FILE_ERROR)
* Encountered end of file
* Closing connection 0
bash-4.2# curl  https://dynamodb.eu-west-1.amazonaws.com -vs
* About to connect() to dynamodb.eu-west-1.amazonaws.com port 443 (#0)
*   Trying 52.94.25.118...
* Connected to dynamodb.eu-west-1.amazonaws.com (52.94.25.118) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* NSS error -5938 (PR_END_OF_FILE_ERROR)
* Encountered end of file
* Closing connection 0
bash-4.2# curl  https://dynamodb.eu-west-1.amazonaws.com -vs
* About to connect() to dynamodb.eu-west-1.amazonaws.com port 443 (#0)
*   Trying 52.119.244.118...
* Connected to dynamodb.eu-west-1.amazonaws.com (52.119.244.118) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* SSL connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate:
* 	subject: CN=dynamodb.eu-west-1.amazonaws.com
* 	start date: Nov 08 00:00:00 2019 GMT
* 	expire date: Oct 23 12:00:00 2020 GMT
* 	common name: dynamodb.eu-west-1.amazonaws.com
* 	issuer: CN=Amazon,OU=Server CA 1B,O=Amazon,C=US
> GET / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: dynamodb.eu-west-1.amazonaws.com
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: Server
< Date: Mon, 27 Jan 2020 11:49:35 GMT
< Content-Type: text/plain
< Content-Length: 42
< Connection: keep-alive
< x-amzn-RequestId: UPGTKK90SV545RRJOGAQB5RQ9RVV4KQNSO5AEMVJF66Q9ASUAAJG
< x-amz-crc32: 3373418776
< 
* Connection #0 to host dynamodb.eu-west-1.amazonaws.com left intact
```

**Comment by eperdeme** [comment]
Weird.
wget always works, curl works intermittent, as does yum.

**Comment by eperdeme** [comment]
Upgraded curl....fails every-time now. curl 7.68.0 (x86_64-redhat-linux-gnu) libcurl/7.68.0 NSS/3.44 zlib/1.2.7 libpsl/0.7.0 (+libicu/50.1.2) libssh2/1.9.0 nghttp2/1.31.1

Swap HTTPS -> TCP works fine.

**Comment by yxue** [comment]
From you yaml configuration, you are *NOT* using protocol sniffing for outbound as you define the port name as "https" and "tcp". 

The problem is 
```
2020-01-27T11:10:33.18320854Z [Envoy (Epoch 0)] [2020-01-27 11:10:33.183][30][debug][conn_handler] [external/envoy/source/server/connection_handler_impl.cc:179] listener filter times out after 100 ms
```

The TLS inspector cannot finish the tls inspection within 100ms and fallback to a default filter chain. This PR https://github.com/istio/istio/pull/20248/files should fix the problem. Or you can disable the protocolDetectionTimeout for outbound by setting `global.proxy.protocolDetectionTimeout=0s`. 

The reason why changing the port name to TCP works is that if the port name is defined as "tcp", there is no tls inspector and no TLS inspector invokes.

**Comment by eperdeme** [comment]
I agree with you @yxue but I've got the following config deployed;

```
---
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  profile: sds
  values:
    global:
      proxy:
        protocolDetectionTimeout: 0s
```

I can see "protocolDetectionTimeout":"0s" in istioctl manifest generate -f dev.yaml . I do an apply, delete the pod but issue stays the same. I'll just wait for 1.5 to appear to fix it. :)

**Comment by yxue** [comment]
@eperdeme could you share the config dump with me? It would help to check the timeout.

**Comment by eperdeme** [comment]
Sure, The proxy-config for the pod is;

https://gist.github.com/eperdeme/8bab478f81311d4378617a1943b921a2

So the dev.yaml profile contains the above protocolDetectionTimeout: 0s".
I've then to be 100% sure

dev.yaml
```
---
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  profile: sds
  telemetry:
    enabled: true
  values:
    global:
      proxy:
        protocolDetectionTimeout: 0s
      outboundTrafficPolicy:
        mode: REGISTRY_ONLY
    pilot:
      enableProtocolSniffingForOutbound: false
      enableProtocolSniffingForInbound: false
    kiali:
      createDemoSecret: true
      enabled: true
      hub: docker-registry.prod.mycompany.plc/mirrors/quay.io/kiali
    tracing:
      enabled: true
      provider: zipkin
    nodeagent:
      image: node-agent-k8s
      env:
        CA_PROVIDER: "Citadel"
        CA_ADDR: "istio-citadel:8060"
        VALID_TOKEN: true
    gateways:
      enabled: true
      istio-ingressgateway:
        enabled: true
        autoscaleMin: 2
        autoscaleMax: 6
        ports:
    # Note that AWS ELB will by default perform health checks on the first port
    # on this list. Setting this to the health check port will ensure that health
    # checks always work. https://github.com/istio/istio/issues/12503
        - port: 15020
          targetPort: 15020
          name: http-status-port
        - port: 80
          targetPort: 80
          name: http2
        - port: 443
          name: https
        # GRPC ingress
        - port: 31400
          name: grpc
          nodePort: 31400
        loadBalancerSourceRanges:
          - "10.0.0.0/8"
          - "100.72.0.0/16"
          - "192.168.2.0/23"
        serviceAnnotations:
          service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0
    sidecarInjectorWebhook:
      enabled: true
      # If true, webhook or istioctl injector will rewrite PodSpec for liveness
      # health check to redirect request to sidecar. This makes liveness check work
      # even when mTLS is enabled.
      rewriteAppHTTPProbe: true
```

```
istioctl manifest apply -f dev.yaml 
kubectl rollout restart deployment --namespace nflynn
```

**Comment by yxue** [comment]
Hmm, it's weird. The config dump shows the listener filter timeout is still 100ms. 
https://gist.github.com/eperdeme/8bab478f81311d4378617a1943b921a2#file-gistfile1-txt-L2247-L2249

**Comment by rmichela** [comment]
@yxue @eperdeme Was this issue fixed in 1.5?

**Comment by howardjohn** [comment]
I believe this is fixed by #24763
